[DataGrid] Data source with editing (#16045)
mui/mui-x: docs/data/data-grid/server-side-data/index.md

```diff
 const dataSource: GridDataSource = {
  getRows: async (params: GridGetRowsParams) => {
    // fetch rows from the server
  },
+ updateRow: async (params: GridUpdateRowParams) => {
+   // update row on the server
+ },
 }
```

{{"demo": "ServerSideEditing.js", "bg": "inline"}}

:::warning
When using the `updateRow()` method, be aware that the Data Source cache is automatically cleared after successful updates to prevent displaying outdated data.
This means any previously cached data will be refetched on the next request.

For applications requiring caching with editing operations, consider implementing server-side cache instead.

If you have a specific use case that requires preserving the client-side cache during edit operations, please open an issue on [GitHub](https://github.com/mui/mui-x/issues/new/choose) to help us understand your requirements.
:::

:::warning
The position of the edited row on the current page is kept the same for UX purposes, even if some features like sorting or filtering are enabled which might affect the row's new position.

This automatically gets fixed when the page is fetched again, since the cache is cleared after the update.

> However, if you have a use-case that requires refetching the current page after editing, please open an issue on [GitHub](https://github.com/mui/mui-x/issues/new/choose) with details about your use-case.

Author: I wanted to cover those cases that are not supported by `fetchRows`, for example re-fetching the expanded row grouping groups or refetching a specific lazy-loading rows range. However, I agree to be simple in explanation here and wait for users' feedback, updated the text: https://github.com/mui/mui-x/pull/16045/commits/724c46ca86ec6430db7b1bbbfa569b6ff9f0c1f6

This method returns a promise that resolves when the row is updated.
If the promise resolves, the grid updates the row and mutates the cache. In case of an error, `onDataSourceError` is triggered with the error object containing the params as mentioned in the [Error handling](#error-handling) section.

```diff
 const dataSource: GridDataSource = {
  getRows: async (params: GridGetRowsParams) => {
    // fetch rows from the server
  },
+ updateRow: async (params: GridUpdateRowParams) => {
+   // update row on the server
+ },
 }
```

{{"demo": "ServerSideEditing.js", "bg": "inline"}}

:::warning
When using the `updateRow()` method, the Data Source cache is automatically cleared after successful updates to prevent displaying outdated data.
This means any previously cached data will be refetched on the next request.

For applications requiring caching with editing operations, consider implementing server-side cache instead.

If you have a specific use case that requires preserving the client-side cache during edit operations, please open an issue on [GitHub](https://github.com/mui/mui-x/issues/new/choose) to help us understand your requirements.
:::

:::warning

> The position and/or visibility of the edited row on the current page is kept, even if features like sorting or filtering are enabled and should affect the row after the value update.

Reviewer: ```suggestion
The position and visibility of the edited row on the current page are maintained—even if features like sorting or filtering are enabled—and will take affect on the row after the values update.
```

> Change in the position and/or visibility is applied when the page is fetched again.

Reviewer: ```suggestion
Any changes to the position or visibility will be applied when the page is fetched again.
```

> Trigger the re-fetch manually by calling `dataSource.fetchRows()` API method.

Reviewer: ```suggestion
You can manually trigger a refetch by calling the `dataSource.fetchRows()` API method.
```

> You could manually trigger the re-fetch by calling the API method `dataSource.fetchRows()`.

Reviewer 2: ```suggestion
Trigger the re-fetch manually by calling `dataSource.fetchRows()` API method.
```

:::

## Error handling

> You could handle the errors with the data source by providing an error handler function using the `onDataSourceError()`.

Reviewer: ```suggestion
You can handle errors with the data source by providing an error handler function with `onDataSourceError()`.
```

> It will be called whenever there's an error in fetching or updating the data.

Reviewer: ```suggestion
This gets called whenever there's an error in fetching or updating the data.
```

This function recieves an error object of type `GridGetRowsError | GridUpdateRowError`.

> For different error types, different `error.params` type is passed as an argument to the callback:

Reviewer: ```suggestion
Each error type has a corresponding `error.params` type which is passed as an argument to the callback:
```
